sklearn/preprocessing/data.py

```diff
@@ -47,41 +50,89 @@
+    function. Both must be initilized to 0.0. In case no scaling is required
+    (caller has with_std=False), last_var can be None. The mean is always
+    required and returned (even if caller has with_std=True), because it is
+    necessary for the calculation of the var. last_n is the number of samples
+    encountered until now.
     """
-    X = np.asarray(X)
-    Xr = np.rollaxis(X, axis)
+    new_X = np.asarray(new_X)
+    new_Xr = np.rollaxis(new_X, axis)
+
+    if last_var is not None:
+        updated_mean, updated_var, updated_n = \
+            _batch_mean_variance_update(new_Xr, last_mean=last_mean,
+                                        last_variance=last_var,
+                                        last_sample_count=last_n)
+    else:
+        new_n = new_Xr.shape[0]
+        updated_n = last_n + new_n
+        updated_mean = (last_mean * last_n + new_Xr.sum(axis=0)) / updated_n
+        updated_var = None
+
+    return updated_mean, updated_var, updated_n
+
+
+def _incremental_min_and_max(new_X, axis=0, last_min=None, last_max=None):
+    """Compute incremental min and max.
 
-    if with_mean:
-        mean_ = Xr.mean(axis=0)
+    last_min, last_max are the statistics computed until now.
+    Both are initilized as None.
+    """
+    new_X = np.asarray(new_X)
+    new_Xr = np.rollaxis(new_X, axis)
+
+    if last_min is not None:
+        updated_min = np.minimum(last_min, np.min(new_Xr, axis))
     else:
-        mean_ = None
+        updated_min = np.min(new_Xr, axis=0)
 
-    if with_std:
-        std_ = Xr.std(axis=0)
-        std_ = _handle_zeros_in_scale(std_)
+    if last_max is not None:
+        updated_max = np.maximum(last_max, np.max(new_Xr, axis))
     else:
-        std_ = None
+        updated_max = np.max(new_Xr, axis=0)
+
+    return updated_min, updated_max
 
-    return mean_, std_
 
+def _incremental_max_abs(new_X, axis=0, last_max_abs=.0):
+    """Compute incremental max absolute value
 
-def _handle_zeros_in_scale(scale):
+    last_max_abs is statistic computed at the last step.
+    It is initialized to .0
+    """
+    new_X = np.asarray(new_X)
+    new_Xr = np.rollaxis(new_X, axis)
+
+    return np.maximum(last_max_abs, np.abs(new_Xr).max(axis=0))
+
+
+def _handle_zeros_in_scale(scale, copy=True):
     ''' Makes sure that whenever scale is zero, we handle it correctly.
 
     This happens in most scalers when we have constant features.'''
 
     # if we are fitting on 1D arrays, scale might be a scalar
     if np.isscalar(scale):
-        if scale == 0:
+        if scale == .0:
             scale = 1.
+        return scale
     elif isinstance(scale, np.ndarray):
-        scale[scale == 0.0] = 1.0
-        scale[~np.isfinite(scale)] = 1.0
-    return scale
+        if copy:
+            # New array to avoid side-effects
+            new_scale = np.array([x for x in scale])
+            new_scale[new_scale == 0.0] = 1.0
+            new_scale[~np.isfinite(new_scale)] = 1.0
```

Comment:
Why silently remove non-finite values? The `check_array`-based input checks in `StandardScaler` and the like should already raise an exception if there are non-finite values in the data. No need to deal with that case again here. If the variance computation overflows (in the event our implementation is not numerically stable) we should not hide it silently.